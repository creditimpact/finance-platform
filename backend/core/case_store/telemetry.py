from typing import Mapping, Any, Optional, Callable
import time

_emit: Optional[Callable[[str, Mapping[str, Any]], None]] = None

def set_emitter(fn: Optional[Callable[[str, Mapping[str, Any]], None]]) -> None:
    """Swap the global emitter for tests or integration."""
    global _emit
    _emit = fn

def emit(event: str, **fields: Any) -> None:
    """Fire-and-forget; safe if no emitter is registered."""
    if _emit:
        try:
            _emit(event, fields)
        except Exception:
            # Never let telemetry break the app
            pass

class timed:
    """Context manager to measure duration_ms and emit after block."""

    def __init__(self, event: str, **base_fields: Any):
        self.event = event
        self.base = base_fields

    def __enter__(self):
        self.t0 = time.perf_counter()
        return self

    def __exit__(self, exc_type, exc, tb):
        dur = (time.perf_counter() - self.t0) * 1000.0
        emit(self.event, duration_ms=round(dur, 3), **self.base)
