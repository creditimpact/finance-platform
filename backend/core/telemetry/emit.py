import json
import logging
from datetime import datetime
from typing import Any, Mapping

logger = logging.getLogger("telemetry")


def emit(event: str, payload: Mapping[str, Any] | None = None) -> None:
    """Emit a structured telemetry event.

    Events are logged as a single JSON line with an ISO timestamp. Failures to
    serialize are logged but otherwise ignored to avoid interrupting the caller.
    """
    record = {"event": event, "ts": datetime.utcnow().isoformat() + "Z"}
    if payload:
        record.update(dict(payload))
    try:
        logger.info(json.dumps(record, sort_keys=True))
    except Exception:
        logger.exception("telemetry_emit_failed")
