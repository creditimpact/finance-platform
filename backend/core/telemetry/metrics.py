import logging
from backend.core.config.flags import FLAGS

log = logging.getLogger(__name__)

def increment(name: str, value: int = 1, tags: dict | None = None) -> None:
    """Increment a counter metric."""
    if not getattr(FLAGS, "metrics_enabled", True):
        return
    log.debug("METRIC increment name=%s value=%s tags=%s", name, value, tags or {})

def gauge(name: str, value: float, tags: dict | None = None) -> None:
    """Record a gauge metric."""
    if not getattr(FLAGS, "metrics_enabled", True):
        return
    log.debug("METRIC gauge     name=%s value=%s tags=%s", name, value, tags or {})
