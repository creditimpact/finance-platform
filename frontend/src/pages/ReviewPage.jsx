import { useLocation, useNavigate } from 'react-router-dom';
import { useState, useEffect } from 'react';
import { submitExplanations, getSummaries } from '../api';

export default function ReviewPage() {
  const { state } = useLocation();
  const navigate = useNavigate();
  const uploadData = state?.uploadData;
  const [explanations, setExplanations] = useState({});
  const [summaries, setSummaries] = useState({});
  const [showSummary, setShowSummary] = useState({});
  const debugEvidence = (() => {
    try {
      return (
        process.env.VITE_DEBUG_EVIDENCE === '1' ||
        new Function('return import.meta.env?.VITE_DEBUG_EVIDENCE')() === '1'
      );
    } catch {
      return false;
    }
  })();

  useEffect(() => {
    if (uploadData?.session_id) {
      (async () => {
        try {
          const res = await getSummaries(uploadData.session_id);
          setSummaries(res?.summaries ?? {});
        } catch (err) {
          console.warn('failed to fetch summaries', err);
          setSummaries({});
        }
      })();
    }
  }, [uploadData?.session_id]);

  if (!uploadData) {
    return <p>No upload data available.</p>;
  }

  const accounts =
    uploadData.accounts?.problem_accounts ?? [
      ...(uploadData.accounts?.negative_accounts ??
        uploadData.accounts?.disputes ??
        []),
      ...(uploadData.accounts?.open_accounts_with_issues ??
        uploadData.accounts?.goodwill ??
        []),
    ];

  // Debug: log first card's props
  if (accounts[0]) {
    console.debug('review-card-props', {
      primary_issue: accounts[0].primary_issue,
      issue_types: accounts[0].issue_types,
      last4: accounts[0].account_number_last4,
      original_creditor: accounts[0].original_creditor,
    });
  }

  const dedupedAccounts = Array.from(
    accounts
      .reduce((map, acc) => {
        const identifier = acc.account_number_last4 ?? acc.account_fingerprint ?? '';
        const key = `${
          acc.normalized_name ?? acc.name?.toLowerCase() ?? ''
        }|${identifier}`;
        const existing = map.get(key);
        if (existing) {
          existing.late_payments = {
            ...(existing.late_payments ?? {}),
            ...(acc.late_payments ?? {}),
          };
          return map;
        }
        map.set(key, acc);
        return map;
      }, new Map())
      .values(),
  );

  const formatIssueType = (type) => {
    switch (type) {
      case 'late_payment':
        return 'Late Payment';
      case 'collection':
        return 'Collection';
      case 'charge_off':
        return 'Charge-Off';
      default:
        return type ? type.charAt(0).toUpperCase() + type.slice(1) : type;
    }
  };

  const handleChange = (key, value) => {
    setExplanations((prev) => ({ ...prev, [key]: value }));
  };

  const handleSubmit = async () => {
    try {
      await submitExplanations({
        session_id: uploadData.session_id,
        filename: uploadData.filename,
        email: uploadData.email,
        explanations,
      });
      navigate('/status');
    } catch (err) {
      console.error(err);
    }
  };

  return (
    <div className="container">
      <h2>Explain Your Situation</h2>
      {dedupedAccounts.map((acc, idx) => {
        const issues = acc.issue_types ?? [];
        const primary = acc.primary_issue;
        const idLast4 = acc.account_number_last4 ?? null;
        const fingerprint = acc.account_fingerprint ?? null;
        const displayId = idLast4 ? `••••${idLast4}` : fingerprint ?? '';
        const secondaryIssues = issues.filter((t) => t !== primary);
        return (
          <div key={idx} className="account-block">
            <p>
              <strong>{acc.name}</strong>
              {displayId && ` ${displayId}`}
              {acc.original_creditor && ` - ${acc.original_creditor}`}
            </p>
            <div className="issue-badges">
              <span className="badge">{primary ? formatIssueType(primary) : 'Unknown'}</span>
              {secondaryIssues.map((type, i) => (
                <span key={i} className="chip">
                  {formatIssueType(type)}
                </span>
              ))}
            </div>
            <textarea
              value={explanations[acc.name] || ''}
              onChange={(e) => handleChange(acc.name, e.target.value)}
              placeholder="Your explanation"
            />
            <small className="helper-text">
              We’ll use your explanation as context to better understand your case. It will
              not be copied word-for-word into your dispute letter.
            </small>
            <div className="summary-toggle">
              <label>
                <input
                  type="checkbox"
                  checked={!!showSummary[acc.account_id]}
                  onChange={(e) =>
                    setShowSummary((prev) => ({ ...prev, [acc.account_id]: e.target.checked }))
                  }
                />
                Show how the system understood your explanation
              </label>
            </div>
            {showSummary[acc.account_id] && summaries[acc.account_id] && (
              <pre className="summary-box">
                {JSON.stringify(summaries[acc.account_id], null, 2)}
              </pre>
            )}
            {debugEvidence && (
              <details className="evidence-toggle">
                <summary>View evidence</summary>
                <pre className="summary-box">
                  {JSON.stringify(
                    {
                      account_trace: acc.account_trace ?? {},
                      bureau_details: acc.bureau_details ?? {},
                    },
                    null,
                    2,
                  )}
                </pre>
              </details>
            )}
          </div>
        );
      })}
      <button onClick={handleSubmit}>Generate Letters</button>
    </div>
  );
}
