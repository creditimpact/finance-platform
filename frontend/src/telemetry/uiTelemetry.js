export const emitUiEvent = (event, payload) => {
  console.debug('telemetry', event, payload);
};
