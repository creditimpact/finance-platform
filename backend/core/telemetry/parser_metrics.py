"""Telemetry helpers for parser audit instrumentation."""

from backend.core.case_store.telemetry import emit


def emit_parser_audit(
    *,
    session_id: str,
    pages_total: int,
    pages_with_text: int,
    pages_empty_text: int,
    extract_text_ms: int,
    call_ai_ms: int | None,
    fields_written: int | None,
    errors: str | None,
) -> None:
    """Emit a ``parser_audit`` telemetry event."""

    emit(
        "parser_audit",
        session_id=session_id,
        pages_total=pages_total,
        pages_with_text=pages_with_text,
        pages_empty_text=pages_empty_text,
        extract_text_ms=extract_text_ms,
        call_ai_ms=call_ai_ms,
        fields_written=fields_written,
        errors=errors,
    )

