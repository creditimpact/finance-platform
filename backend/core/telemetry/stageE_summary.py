from __future__ import annotations

import math
from typing import Any, Mapping, List, Dict, Set

from backend.core.case_store import api as cs_api
from backend.core.case_store.telemetry import emit

_sessions: Dict[str, Dict[str, Any]] = {}


def record_stageA_event(event: str, fields: Mapping[str, Any]) -> None:
    """Record per-account Stage A telemetry for later aggregation.

    Parameters
    ----------
    event:
        Name of the telemetry event. Only ``stageA_eval`` and ``stageA_fallback``
        are recognized.
    fields:
        Telemetry payload emitted for the event.
    """
    session_id = fields.get("session_id")
    if not session_id:
        return
    info = _sessions.setdefault(session_id, {"fallbacks": set(), "latencies": []})
    if event == "stageA_fallback":
        acc_id = fields.get("account_id")
        if acc_id is not None:
            info["fallbacks"].add(str(acc_id))
        latency = fields.get("latency_ms")
        if isinstance(latency, (int, float)):
            info["latencies"].append(float(latency))
    elif event == "stageA_eval":
        latency = fields.get("ai_latency_ms")
        if isinstance(latency, (int, float)):
            info["latencies"].append(float(latency))


def _p95(values: List[float]) -> float:
    if not values:
        return 0.0
    values = sorted(values)
    idx = int(math.ceil(0.95 * len(values))) - 1
    return float(values[idx])


def emit_stageE_summary(session_id: str, problem_accounts: List[Mapping[str, Any]], duration_ms: float | None = None) -> None:
    """Emit aggregated Stage E session summary telemetry."""
    total_accounts = len(cs_api.list_accounts(session_id))
    problem_count = len(problem_accounts or [])
    ai_adoption = sum(1 for acc in problem_accounts if acc.get("decision_source") == "ai")
    sess = _sessions.pop(session_id, {"fallbacks": set(), "latencies": []})
    fallback_count = len(sess.get("fallbacks", set()))
    latencies = sess.get("latencies", [])
    latency_p95 = _p95(latencies)
    emit(
        "stageE_summary",
        session_id=session_id,
        total_accounts=total_accounts,
        problem_accounts=problem_count,
        ai_adoption_pct=ai_adoption / max(1, problem_count),
        fallback_pct=fallback_count / max(1, problem_count),
        ai_latency_p95_ms=latency_p95,
        duration_ms=duration_ms if duration_ms is not None else 0.0,
    )
