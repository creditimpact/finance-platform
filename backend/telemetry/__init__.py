"""Telemetry helper utilities."""

from .metrics import emit_counter

__all__ = ["emit_counter"]
