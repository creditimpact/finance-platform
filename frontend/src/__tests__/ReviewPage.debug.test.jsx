import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import ReviewPage from '../pages/ReviewPage';
import { emitUiEvent } from '../telemetry/uiTelemetry';

jest.mock('../telemetry/uiTelemetry', () => ({
  emitUiEvent: jest.fn(),
}));
jest.mock('../api.ts', () => ({
  API_BASE_CONFIGURED: true,
  API_BASE_URL: 'http://127.0.0.1:5000',
  submitExplanations: jest.fn(),
  getSummaries: jest.fn().mockResolvedValue({ summaries: {} }),
}));

const baseUploadData = {
  session_id: 'sess1',
  filename: 'file.pdf',
  email: 'test@example.com',
};

const account = {
  account_id: 'acc1',
  bureau: 'Equifax',
  name: 'Account 1',
  normalized_name: 'account 1',
  primary_issue: 'late_payment',
  issue_types: ['late_payment'],
  decision_meta: {
    decision_source: 'ai',
    tier: 'Tier1',
    confidence: 0.91,
    fields_used: ['f1', 'f2', 'f3'],
  },
  stage_a_decision: { result: 'ok' },
  problem_reasons: ['late_payment'],
};

beforeEach(() => {
  process.env.REACT_APP_UI_DEV_DEBUG_PANEL = 'true';
  process.env.REACT_APP_UI_SHOW_FIELDS_USED = 'true';
  process.env.REACT_APP_UI_MAX_FIELDS_USED = '2';
  jest.clearAllMocks();
});

test('debug panel shows meta and capped fields', async () => {
  const uploadData = { ...baseUploadData, accounts: { problem_accounts: [account] } };
  const { container } = render(
    <MemoryRouter initialEntries={[{ pathname: '/review', state: { uploadData } }]}> 
      <ReviewPage />
    </MemoryRouter>
  );
  const toggle = await screen.findByText('Debug');
  fireEvent.click(toggle);
  expect(screen.getByText(/source:/)).toHaveTextContent('ai');
  expect(screen.getByText(/tier:/)).toHaveTextContent('Tier1');
  expect(screen.getByText(/confidence:/)).toHaveTextContent('0.91');
  expect(screen.getByText('f1')).toBeInTheDocument();
  expect(screen.getByText('f2')).toBeInTheDocument();
  expect(screen.queryByText('f3')).not.toBeInTheDocument();
  const panel = container.querySelector('.debug-panel');
  expect(panel).toMatchSnapshot();
  const text = panel.textContent;
  expect(text).not.toMatch(/[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/);
  expect(text).not.toMatch(/\b\d{3}-\d{2}-\d{4}\b/);
  expect(text).not.toMatch(/\b\d{3}-\d{3}-\d{4}\b/);
});

test('debug panel hidden when flag off', async () => {
  process.env.REACT_APP_UI_DEV_DEBUG_PANEL = 'false';
  const uploadData = { ...baseUploadData, accounts: { problem_accounts: [account] } };
  render(
    <MemoryRouter initialEntries={[{ pathname: '/review', state: { uploadData } }]}> 
      <ReviewPage />
    </MemoryRouter>
  );
  expect(screen.queryByText('Debug')).not.toBeInTheDocument();
});

test('telemetry fires on expand and collapse', async () => {
  const uploadData = { ...baseUploadData, accounts: { problem_accounts: [account] } };
  render(
    <MemoryRouter initialEntries={[{ pathname: '/review', state: { uploadData } }]}> 
      <ReviewPage />
    </MemoryRouter>
  );
  const toggle = await screen.findByLabelText(/Show how the system understood/i);
  fireEvent.click(toggle);
  expect(emitUiEvent).toHaveBeenCalledWith('ui_review_expand', {
    session_id: 'sess1',
    account_id: 'acc1',
    bureau: 'Equifax',
    decision_source: 'ai',
    tier: 'Tier1',
  });
  fireEvent.click(toggle);
  expect(emitUiEvent).toHaveBeenCalledWith('ui_review_collapse', {
    session_id: 'sess1',
    account_id: 'acc1',
    bureau: 'Equifax',
  });
});

