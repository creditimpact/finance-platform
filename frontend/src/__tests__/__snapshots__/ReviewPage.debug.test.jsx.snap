// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug panel shows meta and capped fields 1`] = `
<details
  class="debug-panel"
  open=""
>
  <summary>
    Debug
  </summary>
  <div
    class="decision-meta"
  >
    <div>
      source: 
      ai
    </div>
    <div>
      tier: 
      Tier1
    </div>
    <div>
      confidence: 
      0.91
    </div>
  </div>
  <div
    class="fields-used"
  >
    <div>
      fields_used:
    </div>
    <ul>
      <li>
        f1
      </li>
      <li>
        f2
      </li>
    </ul>
  </div>
  <pre
    class="debug-json"
  >
    {
  "result": "ok"
}
  </pre>
</details>
`;
